packages/vite/src/node/build.ts

```diff
@@ -502,8 +503,48 @@ async function doBuild(
         }
       })
 
-      // stop watching
-      watcher.close()
+      if (config.configFile) {
+        const configWatcher = chokidar.watch(config.configFile, {
+          ignoreInitial: true
+        })
+        type Watcher = RollupWatcher & {
+          firstWatcher?: RollupWatcher
+        }
+        const lastWatcher = watcher as Watcher
+        lastWatcher.once('close', () => {
+          configWatcher.close()
+        })
+        configWatcher.once('change', () => {
+          lastWatcher.close()
```

Comment:
Couldn't re-starting the config file watcher end up missing change events? If the user saves the config file several times in a row, it may miss the change while it is restarting the build, no?

I think that a more robust solution would be to leave doBuild untouched and create the chokidar file watcher only once, closing it once every rollup watcher is closed. 
Looks like we are also returning the rollup watcher that will end up being outdated. Should we return the chokidar watcher if there is a config file? Since we are going to be re-starting it, we shouldn't leak the rollup watcher in this case.

I'm thinking of something like this outside of doBuild (or wrapping it). Warning: untested code.
```js
if (config.configFile) {
  const configWatcher = chokidar.watch(config.configFile)

  let activeWatchers = []
  const onClose = closedWatcher => {
    activeWatchers = activeWatchers.filter( entry => entry.watcher !== closedWatcher )
    if( activeWatchers.length === 0 ) {
      configWatcher.close()
    }
  }
  const outdateWatcherEntry = entry => {
    entry.watcher?.close()
    entry.outdated = true
  })

  configWatcher.once('close', () => {
    activeWatchers.forEach(outdateWatcherEntry)
  })

  configWatcher.on('change', () => {
    const oldWatchers = [...activeWatchers]
    
    // Push the new watcher entry before the promise to avoid closing the 
    // configWatcher while waiting for reload to finish
    const entry = { watcher: null, outdated: false }
    activeWatchers.push(entry)

    oldWatchers.forEach(outdateWatcherEntry)
    
    // Reload the config and create another watcher
    const watcherPromise = doBuild(inlineConfig)
    
    watcherPromise.then( watcher => {
      entry.watcher = watcher
      watcher.once('close', () => onClose(watcher))
      if( entry.outdated ) {
        watcher.close()
      }
    })
    
    watcherPromise.catch((error) => {
      process.emit('uncaughtException', error)
    })
  })

  return configWatcher
}
else {
  return doBuild(inlineConfig)
}
```

Reply:
> Couldn't re-starting the config file watcher end up missing change events? If the user saves the config file several times in a row, it may miss the change while it is restarting the build, no?

I fixed this by comparing `configFile` mtime before/after it's been loaded. I did some manual testing, and it's catching changes between restarts now. https://github.com/vitejs/vite/pull/4272/commits/4d90f201e437b4b5191968dde5cf8d02dcd95475

> Looks like we are also returning the rollup watcher that will end up being outdated.

We *could* use a `Proxy` to redirect method calls to the latest Rollup watcher. I thought it was "good enough" to just override its `close` method instead. If you think otherwise, you could write a follow-up PR. 👍

Reply:
Actually, I'll see about the proxying bit. It shouldn't be difficult.

Reply:
Ya, I think it is better to avoid exposing the rollup watcher. Users may think they can subscribe to events and these will be lost once the config file changes.

Reply:
We *could* forward Rollup events from the last watcher to the first watcher. :)

Reply:
Ha, I don't know enough about use cases for having these exposes to justify the complexity. But I don't think the current API returning the watcher was added because it was needed, it was more because it was there at hand to return it. So this could be done when there is a use case (maybe from Electron or Tauri?)

Reply:
In that case, this PR is ready to merge. If anyone tries to use the Rollup watch events, they'll open an issue when it doesn't work.

Reply:
For reference, is there a reason for the current design where you need to manually check the config file `mtime`? Wouldn't it be more robust to create the chokidar watcher only once and keep it up until the end?

Reply:
They're equal in robustness. The "problem" with your approach (besides my opinion that it's more complex) is that the chokidar watcher is recreated on each `doBuild` call, so we would need another `doBuildWatch` function to ensure the chokidar watcher is created just once. Additionally, returning the Rollup watcher keeps the door open for the ability to observe Rollup events in the future (when developers need it enough to open an issue about it).

Reply:
Yes, at least myself I don't see having that extra wrapper for doBuild as an issue. But ya, if this works then it's fine, I let others check it out. We'll need approval from Evan anyway (although maybe for this particular feature it may be merged with enough approvals as there aren't any new config)

Reply:
Yeah, I'm about 99.99% certain Evan wouldn't object to this PR. :P

This PR is a little difficult to manually test. I added logs to ensure the `configFileLastModified` check was being used when I saved twice fast enough. I can guarantee it works, so if y'all would rather avoid the hassle, we can just merge it.